### PR TITLE
Unit tests for levels and roles

### PR DESCRIPTION
**Pull request by ice-nestor**
In this pr, unit tests are added for levels and roles.

### CODE REVIEW

File: cmd/santa-sleigh/santa_sleigh.go [hunk 1]
```diff
@@ -41,9 +39,9 @@ func (s *service) RegisterRoutes(router *server.Router) {
 }
 
 func (s *service) Init(ctx context.Context, cancel context.CancelFunc) {
-	s.tasksProcessor = tasks.StartProcessor(ctx, cancel)
+	// s.tasksProcessor = tasks.StartProcessor(ctx, cancel)
```
Comment (ice-ares) [inline cmd/santa-sleigh/santa_sleigh.go:42]: rollback

File: levels-and-roles/completed_levels_test.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+		},
+	}
+
+	for _, tt := range testCases { //nolint:gocritic,revive // it's a test, no need for optimization
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			got := tt.p.reEvaluateCompletedLevels(tt.repo)
+			require.Equal(t, tt.expected, got)
```
Comment (ice-ares) [inline levels-and-roles/completed_levels_test.go:191]: equal values

File: levels-and-roles/completed_levels.go [hunk 3]
```diff
@@ -347,7 +347,7 @@ func (*userTableSource) newlyAddedAgendaContacts(us *users.UserSnapshot) map[str
 outer:
 	for _, afterAgendaPhoneNumberHash := range after {
 		if afterAgendaPhoneNumberHash == "" || strings.EqualFold(afterAgendaPhoneNumberHash, us.User.PhoneNumberHash) {
-			continue outer
+			continue
```
[the remark this answers is not in the document]
Comment (ice-nestor) [reply levels-and-roles/completed_levels.go]: yeah, but there is no need to specify label

File: levels-and-roles/completed_levels_test.go [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+			}},
+			expected: &users.Enum[LevelType]{Level1Type},
+		},
+		{
+			name:     "no levels are completed, return nil",
+			p:        &progress{},
+			repo:     &repository{cfg: &config{}},
+			expected: nil,
+		},
+	}
```
Comment (ice-cronus) [inline levels-and-roles/completed_levels_test.go:183]: create complex test with achieving levels from different groups as well

Comment (ice-cronus) [reply levels-and-roles/completed_levels_test.go:183]: also achieving next level (when user has level1 and passed requirments for level2 in the same group) is missing too

Comment (ice-cronus) [reply levels-and-roles/completed_levels_test.go:183]: also when users achieves multiple levels at once

**Review by ice-cronus — CHANGES_REQUESTED** [review]

File: levels-and-roles/levels_and_roles_test.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+			actual:         &users.Enum[LevelType]{Level1Type, Level2Type, Level3Type},
+			expectedSubset: []LevelType{Level1Type, Level2Type, Level3Type},
+			expected:       true,
+		},
+		{
+			name:           "actual levels are completed, except one",
+			actual:         &users.Enum[LevelType]{Level1Type, Level2Type, Level3Type},
+			expectedSubset: []LevelType{Level1Type, Level2Type},
+			expected:       true,
+		},
```
Comment (ice-myles) [inline levels-and-roles/levels_and_roles_test.go:47]: No 1 case when
```
if !completed {
    return false
}
```
will be called.
It's when you have expected and actual, but they are different.

**Review by ice-myles — CHANGES_REQUESTED** [review]

File: levels-and-roles/completed_levels.go [hunk 6]
```diff
@@ -51,7 +51,7 @@ func (s *miningSessionSource) upsertProgress(ctx context.Context, miningStreak u
 	if pr, err := s.getProgress(ctx, userID); err != nil && !errors.Is(err, storage.ErrRelationNotFound) ||
 		(pr != nil && pr.CompletedLevels != nil &&
 			(len(*pr.CompletedLevels) == len(&AllLevelTypes) ||
-				AreLevelsCompleted(pr.CompletedLevels, Level1Type, Level2Type, Level3Type, Level4Type, Level5Type))) {
+				areLevelsCompleted(pr.CompletedLevels, Level1Type, Level2Type, Level3Type, Level4Type, Level5Type))) {
```
[the remark this answers is not in the document]
Comment (ice-nestor) [reply levels-and-roles/completed_levels.go]: done

File: cmd/santa-sleigh/santa_sleigh.go (hunk 1 above)
Comment (ice-nestor) [reply cmd/santa-sleigh/santa_sleigh.go]: done

File: levels-and-roles/levels_and_roles_test.go (hunk 5 above)
Comment (ice-nestor) [reply levels-and-roles/levels_and_roles_test.go:47]: added

File: levels-and-roles/completed_levels_test.go (hunk 4 above)
Comment (ice-nestor) [reply levels-and-roles/completed_levels_test.go]: added, please re-check

File: levels-and-roles/completed_levels_test.go [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+				User: &users.User{
+					PublicUserInformation:   users.PublicUserInformation{ID: "bogus"},
+					AgendaPhoneNumberHashes: "Ef86A6021afCDe5673511376B1,Ef86A6021afCDe5673511376B2,,Ef86A6021afCDe5673511376B4",
+				},
+				Before: &users.User{
+					PublicUserInformation:   users.PublicUserInformation{ID: "bogus"},
+					AgendaPhoneNumberHashes: "Ef86A6021afCDe5673511376B1,Ef86A6021afCDe5673511376B2",
+				},
+			},
+			expected: map[string]struct{}{
```
[the remark this answers is not in the document]
Comment (ice-nestor) [reply levels-and-roles/completed_levels_test.go]: added

**Pull request merged**